Page document and frame loops should traverse from the remote main frame

charliewolfe opened the pull request:
#### 86a15eb9f4a9b00b7fa0e141e57d643a7316b8bf
<pre>
Page document and frame loops should traverse from the remote main frame
<a href="https://bugs.webkit.org/show_bug.cgi?id=262513">https://bugs.webkit.org/show_bug.cgi?id=262513</a>
rdar://116374168

Reviewed by Alex Christensen.

Several things rely on the document and frame loops in `WebCore::Page` which begin traversing from the
main frame. With site isolation, the main frame may be in a different process, so the frame traversal
will never happen. We should do this traversal even if the main frame is remote.

Also rename forEachFrame to forEachLocalFrame.

* Source/WebCore/inspector/agents/InspectorApplicationCacheAgent.cpp:
(WebCore::InspectorApplicationCacheAgent::getFramesWithManifests):
* Source/WebCore/inspector/agents/page/PageRuntimeAgent.cpp:
(WebCore::PageRuntimeAgent::reportExecutionContextCreation):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::~Page):
(WebCore::Page::notifyToInjectUserScripts):
(WebCore::Page::forEachDocumentFromMainFrame):
(WebCore::Page::forEachDocument const):
(WebCore::Page::forEachLocalFrame):
(WebCore::Page::forEachFrame): Deleted.
* Source/WebCore/page/Page.h:

Canonical link: <a href="https://commits.webkit.org/268798@main">https://commits.webkit.org/268798@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fa0d77f40d3e51602f98c113db084b5a31645873

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20695 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/21112 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21769 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22590 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19317 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20932 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24367 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21295 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/20644 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20917 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20741 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17993 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23446 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/17906 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18799 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/25070 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18985 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/18977 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/23017 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19566 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16625 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18787 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23124 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2555 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19384 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 reviewed (changes requested):
This isn't correct because a root frame can be a grandchild frame, which would traverse it twice.  Instead, we need to pass a Frame& to Page::forEachDocumentFromMainFrame.  forEachFrame could be renamed to forEachLocalFrame, but should otherwise remain unchanged.  forEachWindowEventLoop should remain unchanged.

charliewolfe committed:
Page document and frame loops should traverse from the remote main frame
https://bugs.webkit.org/show_bug.cgi?id=262513
rdar://116374168

Reviewed by Alex Christensen.

Several things rely on the document and frame loops in `WebCore::Page` which begin traversing from the
main frame. With site isolation, the main frame may be in a different process, so the frame traversal
will never happen. We should do this traversal even if the main frame is remote.

Also rename forEachFrame to forEachLocalFrame.

* Source/WebCore/inspector/agents/InspectorApplicationCacheAgent.cpp:
(WebCore::InspectorApplicationCacheAgent::getFramesWithManifests):
* Source/WebCore/inspector/agents/page/PageRuntimeAgent.cpp:
(WebCore::PageRuntimeAgent::reportExecutionContextCreation):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::~Page):
(WebCore::Page::notifyToInjectUserScripts):
(WebCore::Page::forEachDocumentFromMainFrame):
(WebCore::Page::forEachDocument const):
(WebCore::Page::forEachLocalFrame):
(WebCore::Page::forEachFrame): Deleted.
* Source/WebCore/page/Page.h:

Canonical link: https://commits.webkit.org/268798@main
